Policy Violations Page Blog (#12588)
pulumi/docs: content/blog/centralized-policy-violations/index.md

---
title: "Introducing the Centralized Policy Violations Page: Streamlined Compliance Management"
allow_long_title: true
date: 2024-08-26T09:00:00-07:00
draft: false
meta_desc: The new Centralized Policy Violations page in Pulumi offers enhanced visibility and control by consolidating all policy violations.
meta_image: meta.png
authors:
    - meagan-cojocar
tags:
    - releases
    - policy
---

> We’re excited to introduce a new page in Pulumi Cloud designed to simplify policy management and compliance: the centralized Policy Violations page. This powerful addition provides a single, comprehensive view of all policy violations across your organization, helping you maintain control and visibility over your infrastructure.

Reviewer: ```suggestion
We’re excited to introduce a new centralized Policy Violations destination in Pulumi Cloud to simplify policy management and compliance. This powerful addition provides a comprehensive view of all policy violations across your organization, helping you maintain visibility and control over your infrastructure.
```
This is just opinion of course but I found the original a bit hard to read.

Pulumi’s Policy as Code (PaC) feature, [CrossGuard](/docs/using-pulumi/crossguard), allows you to define and enforce policies for your infrastructure using familiar programming languages. By integrating these policies directly into your infrastructure code, you can automate compliance checks, gate deployments to ensure security best practices, and maintain governance across all your providers.

**Note:** Pulumi CrossGuard is available in the Pulumi Business Critical edition.

## Centralized Policy Violations: Why It Matters

> In large and complex environments, ensuring compliance across multiple projects and stacks can be challenging. The new Policy Violations page addresses this challenge by consolidating all violations into one place, allowing you to quickly assess the overall compliance of your infrastructure and separate signal from noise. This streamlined view not only saves time but also reduces the risk of missing critical issues.

Reviewer: ```suggestion
Ensuring compliance across multiple projects and stacks can be challenging in large and complex environments. The new Policy Violations page addresses this challenge by consolidating all violations into one place. This allows you to quickly assess the overall compliance of your infrastructure and separate signal from noise. This streamlined view saves time and reduces the risk of missing critical issues.
```

By having all violations on a single page, organization admins can efficiently prioritize remediation efforts and identify patterns in non-compliance. They can triage the violations and send the critical ones to teams to resolve.

{{< video title="Centralized Policy Violations Overview" src="policy-violations-overview.mp4" autoplay="false" loop="false" >}}

## Grouping Violations: Enhanced Insights Across Multiple Dimensions

> One of the standout features of the new page is the ability to group violations by various fields, including Policy Pack, Policy, Project, Stack, Resource, Reason, Enforcement Level, Version, and Violation Date. This flexibility allows you to gain deeper insights into your infrastructure’s compliance landscape.

Reviewer 2: You currently can not group by reason, violation date, version, or update. I can change this if youd like for all of them but the violation date might look odd as the grouping would be down to the exact timestamp (we could update this to be by the string groupings but it might take some time to get right).

Author: Any reason why we prevent grouping by those?

Reviewer 2: The cardinality on the reason one will be very high and be a very expensive query when we go to the backend because of how long the text is (it has to text match on all of those rows to build the results). The version and update was just from my initial thoughts so I can change those pretty easily. The violation date would need some more work to get right as the actual data is down to the timestamp where as we show a "rounded" string in the grid. The grouping would be based on that timestamp value so we would need to customize it to group by the strings instead.

Author: Reason maps with Policy 1:1 I believe, so I don't know that it should be treated separately. If its possible, I think I would make everything groupable except Violation Date, similar to the filters decision we made. This isn't urgent, but if we are going to do it I might leave in the blog and we do it post-launch.

Reviewer: I agree with Meagan. Reason actually has very low cardinality.  The main challenge for grouping by reason is just that i t is ugly as a group.  Ideally if they grouped by Reason we would group by Policy instead and explain why, but that might be an awkward UX.

Reviewer 2: Ok I can change them to allow grouping on all fields but Violation Date. I have some concerns about backend performance down the road but the data if the data size grows a lot, it is small right now.

> One of the standout features of the new page is the ability to group violations by various fields, including Policy Pack, Policy, Project, Stack, Resource, Reason, Enforcement Level, Version, and Violation Date. This flexibility allows you to gain deeper insights into your infrastructure’s compliance landscape.

Reviewer: Do we want to actually list all the fields you can group on, as this will likely evolve?  Should we just list a few with words like "to name a few" to indicate there are more?

> One of the standout features of the new page is the ability to group violations by various fields, including Policy Pack, Policy, Project, Stack, Resource, Reason, Enforcement Level, Version, and Violation Date. This flexibility allows you to gain deeper insights into your infrastructure’s compliance landscape.

Author: ```suggestion
One of the standout features of the new page is the ability to group violations by various fields, including Policy Pack, Policy, Project, Stack, Resource, and more! This flexibility allows you to gain deeper insights into your infrastructure’s compliance landscape.
```

Grouping violations based on these fields enables you to answer critical questions such as:

> - **Policy Pack and Policy**: Which Policy Packs and the Policies within them are most frequently violated?

Reviewer: ```suggestion
- **Policy Pack and Policy**: Which Policy Packs and Policies have the most frequent violations?
```

- **Project and Stack**: Are there specific projects or stacks that are more prone to compliance issues?

> - **Resource**: Which resources are consistently failing policies?

Reviewer: ```suggestion
- **Resource**: Which resources have failed policies?
```

- **Reason and Enforcement Level**: What are the common reasons for violations, and how critical are they?
- **Violation Date**: When are violations most likely to occur, and are there patterns over time?

These insights empower you to refine your policies, address the most problematic areas, and improve your overall compliance strategy.

## Accessing Policy Violations via API

> To further enhance your compliance management, you can leverage the Pulumi Cloud REST API to programmatically access policy violations, trigger automated workflows, and import data into your analytics systems. By using the API, you can automate the monitoring of your infrastructure’s compliance status, ensuring that policy violations are promptly detected and addressed. This capability enables you to integrate Pulumi's compliance data with your existing tools, creating a seamless and efficient compliance management process.

Reviewer: ```suggestion
To further enhance your compliance management, you can leverage the Pulumi Cloud REST API to programmatically access policy violations, trigger automated workflows, and import data into your analytics systems. The API can automate the monitoring of your infrastructure’s compliance status, ensuring that policy violations are promptly detected and addressed. This capability enables you to integrate Pulumi's compliance data with your existing tools, creating a seamless and efficient compliance management process.
```

For example, you can use the following command to retrieve a list of policy violations:

```bash
curl \
  -H "Accept: application/vnd.pulumi+8" \
  -H "Content-Type: application/json" \
  -H "Authorization: token $PULUMI_ACCESS_TOKEN" \
  https://api.pulumi.com/api/orgs/{organization}/policyresults/violations
```

For more details on how to integrate this API into your workflow, visit the [Pulumi API documentation](/docs/pulumi-cloud/cloud-rest-api/#list-policy-violations).

## Streamlining Compliance with Future Enhancements

> Centralizing and surfacing policy violations is just the beginning. Future updates will include additional capabilities like advanced search and filtering, and making it even easier to manage and resolve violations. We’re committed to continuing to provide the tools you need to keep your infrastructure secure, compliant, and running smoothly.

Reviewer: ```suggestion
Centralizing and surfacing policy violations is just the beginning. Future updates will include additional capabilities like advanced search and filtering, and making it even easier to manage violations. We’re committed to providing the tools you need to keep your infrastructure secure, compliant, and running smoothly.
```

## Wrapping It Up

> The addition of the Centralized Policy Violations Page to Pulumi is a significant step forward in simplifying policy management and enhancing compliance. By bringing all violations into a single, easily accessible view, we’re making it easier for admins to maintain control and ensure that their infrastructure remains secure and compliant.

Reviewer: ```suggestion
The Centralized Policy Violations Page to Pulumi is a significant step forward in simplifying policy management and enhancing compliance. By bringing all violations into a single, easily accessible view, we’re making it easier for admins to maintain control and ensure their infrastructure remains secure and compliant.
```
